feat(recipes): add recipe for local first data using powersync

trevor-coleman opened the pull request:
Adds a recipe for using PowerSync and Supabase to sync local and backend data.

Topcis include:

* Installing the SDK and Dependencies
* Authenticating with Supabase
* Connecting PowerSync to Supabase
* Fetching and Managing TodoLists
* Sharing the Database With Screens
* Fetching and Managing Todos

frankcalise commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,713 @@
+  keys**.
+
+2. **Configure or disable email verification in your Supabase project.**
+  - by default, Supabase requires email verification for new users. This should be configured for any production apps.
```
`by` => `By`

frankcalise commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,713 @@
+## Prerequisites
+
+1. **An Ignite app using `Expo CNG` or `Bare` workflow**
+   - PowerSync requires native modules, so you cannot use Expo Go
```
We can give a copy/pastable command to generate this real quickly

```
```bash
npx ignite-cli@latest new PizzaSyncApp --remove-demo --workflow=cng --yes
cd PizzaApp
```

frankcalise commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,713 @@
+/** @type {import('@babel/core').TransformOptions['plugins']} */
+const plugins = [
+        //... other plugins
+        '@babel/plugin-transform-async-generator-functions',  // <-- Add this
```
You can put `// success-line` on the line before this one so that it highlights it green as an addition and calls it out a bit better

frankcalise reviewed (commented):
Overall looks good, nice job as it's a pretty large topic to cover!

Left some comments that can maybe help spruce a few things up.

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,713 @@
+} from "@journeyapps/powersync-sdk-react-native";
+import {createClient} from "@supabase/supabase-js";
+import Config from "app/config";
+import {KVStorage} from "./kv-storage";
```
~Am I supposed to grab this from somewhere?~

The link above was wrong. Now what's interesting is that I also need to install `expo-secure-storage` to make this work.

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,713 @@
+import React, {PropsWithChildren} from "react";
+import {AbstractPowerSyncDatabase, RNQSPowerSyncDatabaseOpenFactory} from '@journeyapps/powersync-sdk-react-native';
+import {SupabaseConnector, supabaseConnector} from "./supabase"; // Adjust the path as needed
+import {AppSchema} from './schema'; // Adjust the path as needed
```
What does this reference?

morganick replied:
This is due to a typo above where it's exported as `TodoSchema`.

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,713 @@
+Persisting the Supabase session is essential for maintaining user sessions across app restarts. Unlike web environments
+where `localStorage` is available, React Native requires a different approach for secure storage.
+
+The`react-native-supabase-todolist` demo from PowerSync has a [simple implementation of a kv-store](https://github.com/powersync-ja/powersync-js/blob/main/demos/react-native-supabase-todolist/library/storage/SupabaseStorageAdapter.ts)
```
This is not the file you're looking for.

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,713 @@
+
+### Initializing the PowerSync Instance, and providing it to the app
+
+We need a sinple point of connection to the PowerSync instance that we can interact with throughout the app. 
```
```suggestion
We need a single point of connection to the PowerSync instance that we can interact with throughout the app. 
```

morganick reviewed (changes requested):
This is as far as I made it today. I'll wrap the review up in the morning.

trevor-coleman commented:
Appreciate the comments -- it's hard to keep everything consistent because I have a working app, and then make changes to make the writing simpler and it's hard to keep them both in sync.

trevor-coleman commented on docs/recipes/LocalFirstDataWithPowerSync.md:
````diff
@@ -0,0 +1,713 @@
+uploadData: (database: AbstractPowerSyncDatabase) => Promise<void>;
+```
+
+The Powersync docs provide [detailed instructions](https://docs.powersync.com/client-sdk-references/react-native-and-expo#id-2.-create-a-backend-connector) on how to implement the `PowerSyncBackendConnector` interface.
````
hrm yeah I'll move it to a section for "other backends"

trevor-coleman commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,713 @@
+import 'react-native-polyfill-globals/auto';
+import React, {PropsWithChildren} from "react";
+import {AbstractPowerSyncDatabase, RNQSPowerSyncDatabaseOpenFactory} from '@journeyapps/powersync-sdk-react-native';
+import {SupabaseConnector, supabaseConnector} from "./supabase"; // Adjust the path as needed
```
Ah missed they type export:

```
export interface SupabaseConnector extends PowerSyncBackendConnector {
    fetchCredentials: typeof fetchCredentials
    uploadData: typeof uploadData
}
```

trevor-coleman replied:
Added this in

trevor-coleman commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,713 @@
+    }, []);
+
+    return (
+        <DatabaseContext.Provider value={database}>
```
The file should be `.tsx` that's probably the problem.

Also use this implementation:

```tsx
import React, {PropsWithChildren} from "react";
import {AbstractPowerSyncDatabase, RNQSPowerSyncDatabaseOpenFactory} from '@journeyapps/powersync-sdk-react-native';
import {SupabaseConnector, supabaseConnector} from "./supabase"; // Adjust the path as needed
import {AppSchema} from './schema'; // Adjust the path as needed

export class Database {
    // We expose the PowerSync instance and the Supabase connector for easy access elsewhere in the app
    powersync: AbstractPowerSyncDatabase;
    supabase: SupabaseConnector;

    /**
     * Initialize the Database class with a new PowerSync instance
     */
    constructor() {
        const factory = new RNQSPowerSyncDatabaseOpenFactory({
            schema: AppSchema,
            dbFilename: 'sqlite.db',
        });
        this.powersync = factory.getInstance();
        // include the Supabase connector for easy access
        this.supabase = supabaseConnector;
    }

    /**
     * Initialize the PowerSync instance and connect it to the Supabase backend.
     * This will call `fetchCredentials` on the Supabase connector to get the session token.
     * So if your database requires authentication, the user will need to be signed in before this is called.
     */
    async init() {
        await this.powersync.init();
        // Connect the PowerSync instance to the Supabase backend through the connector
        await this.powersync.connect(supabaseConnector);
    }
}

// Then we create an instance of the Database class and create a context for it to hold a stable reference
export const database = new Database();

export const DatabaseContext = React.createContext<Database | undefined>(undefined);

export const useDatabase = () => {
    const context: Database | undefined = React.useContext(DatabaseContext);
    if (!context) {
        throw new Error("useDatabase must be used within a DatabaseProvider");
    }
    return context;
};

// Finally, we create a provider component that initializes the database and provides it to the app
export function DatabaseProvider({children}: PropsWithChildren<{}>) {
    useEffect(() => {
        database.init().catch(console.error);
    }, []);

    return (
        <DatabaseContext.Provider value={database}>
            <PowerSyncContext.Provider value={database.powersync}>
                {children}
            </PowerSyncContext.Provider>
        </DatabaseContext.Provider>
    );
}
```

trevor-coleman commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,713 @@
+// app/services/database/schema.ts
+import {Schema, Table, Column, ColumnType} from '@journeyapps/powersync-sdk-react-native';
+
+export const todoSchema = new Schema([
```
replaced the schema declaration

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,741 @@
+};
+
+// Finally, we create a provider component that initializes the database and provides it to the app
+export function DatabaseProvider({children}: PropsWithChildren<{}>) {
```
ESLint is complaining about the empty `{}`.
```suggestion
export function DatabaseProvider({children}: { children: ReactNode }) {
```

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,741 @@
+
+    return (
+        <DatabaseContext.Provider value={database}>
+            <PowerSyncContext.Provider value={database.powersync}>
```
This is not imported.

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,741 @@
+import {usePowerSyncQuery} from '@journeyapps/powersync-sdk-react-native';
+
+const StaticItemList = () => {
+    const items = usePowerSyncQuery(`SELECT * FROM items`);
```
We switch from todos to items.

trevor-coleman commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,741 @@
+        token: session.access_token,
+        expiresAt: new Date(session.expires_at * 1000),
+        userID: session.user.id,
+    };
```
Hrm  -- this I basically lifted from their demo. I get the sense that they are Java devs because they do a lot of control flow with exceptions.

I'll add a bit to return null. May as well implement it correctly.

trevor-coleman commented:
Here's what the finished app looks like:

![CleanShot 2024-03-28 at 13 04 33](https://github.com/infinitered/ignite-cookbook/assets/22041394/bce6b529-6f62-4749-9567-8e18288925de)

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
````diff
@@ -0,0 +1,2024 @@
+
+```tsx
+import {Button, Icon, ListItem, Screen, Text, TextField} from "app/components"
+import {SignedInNavigatorProp} from "app/navigators"
````
```suggestion
import {SignedInNavigatorScreenProps} from "app/navigators"
```

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,2024 @@
+import {FlatList, Pressable, TextStyle, View, ViewStyle} from "react-native"
+import {SafeAreaView} from "react-native-safe-area-context"
+
+interface TodoListScreenProps extends SignedInNavigatorProp<"TodoList"> {}
```
```suggestion
interface TodoListScreenProps extends SignedInNavigatorScreenProps<"TodoList"> {}
```

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,2024 @@
+        icon={"x"}
+        onPress={() => handleRemoveTodo(item.id)}/>
+    </View>
+  }, [handleRemoveTodo])
```
The delete button isn't used in the code.

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
````diff
@@ -0,0 +1,2024 @@
+We're going to use expo-checkbox to show the completed state, so we'll need to install it:
+
+```shell
+npx expo add @react-native-community/checkbox
````
Where do we use this?

trevor-coleman replied:
Ah we don't -- i removed it because it was annoying and I could just use strikethrough

morganick reviewed (changes requested):
Here are the things that I ran into. This was a much better experience than the first run through. For some reason it isn't syncing through to Supabase so I'll have to check my configuration there. I can pull data from Supabase into PowerSync, but the application doesn't push anything.

I'm able to create things inside the application. It just stays locally.

trevor-coleman commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,2024 @@
+      Welcome: undefined
+      // highlight-next-line
+      Auth: undefined              // <---REMOVE THIS LINE
+      TodoList: { listId: string } 
```
we handle TodoList and Welcome in the next step

trevor-coleman commented:
Thanks for that @morganick  -- integrated it all and made a few extra tweaks:

1. moved `use-auth` into `/services/database` -- it's the same service (supabase) and it puts all our stuff in one place
2. went through and added comments with the filename to all codeblocks where it makes sense (`// app/path/to/file.tsx`)
3. Made all text references `PowerSync` and all variables `powersync` (this matches their code)
4. replaced `uuid` with `expo-crypto` -- single install instead of two, and it's better and more react-nativey

trevor-coleman commented:
I think we're pretty close -- at least I feel good enough to share it with PowerSync now.

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,2037 @@
+  created_by: string;
+  completed_by?: string;
+  list_id: string;
+  photo_id?: string;
```
This shouldn't be here.

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,2037 @@
+
+  // Sign out the current user
+  const signOut = useCallback(async () => {
+    setLoading(true)
```
We need to disconnect from syncing here. Ref: https://github.com/powersync-ja/powersync-js/blob/cef404919d7c1f700cb779ef430f8d4845221bd4/demos/react-native-supabase-todolist/app/views/signout.tsx#L12

morganick reviewed (changes requested):
In my version, I'm still getting:
```
LOG  PowerSync status changed {"connected": true, "dataFlow": {"downloading": false, "uploading": false}, "lastSyncedAt": 2024-04-02T14:53:18.701Z}
 LOG  PowerSync status changed {"connected": true, "dataFlow": {"downloading": false, "uploading": false}, "lastSyncedAt": 2024-04-02T14:53:18.701Z}
 LOG  PowerSync status changed {"connected": true, "dataFlow": {"downloading": false, "uploading": true}, "lastSyncedAt": 2024-04-02T14:53:18.701Z}
 LOG  PowerSync status changed {"connected": true, "dataFlow": {"downloading": false, "uploading": true}, "lastSyncedAt": 2024-04-02T14:53:18.701Z}
 LOG  PowerSync status changed {"connected": true, "dataFlow": {"downloading": false, "uploading": false}, "lastSyncedAt": 2024-04-02T14:53:18.701Z}
 LOG  PowerSync status changed {"connected": true, "dataFlow": {"downloading": false, "uploading": false}, "lastSyncedAt": 2024-04-02T14:53:18.701Z}
 LOG  PowerSync status changed {"connected": false, "dataFlow": {"downloading": false, "uploading": false}, "lastSyncedAt": 2024-04-02T14:53:18.701Z}
 LOG  PowerSync status changed {"connected": false, "dataFlow": {"downloading": false, "uploading": false}, "lastSyncedAt": 2024-04-02T14:53:18.701Z}
 DEBUG  [PowerSyncStream] Stream input empty
```
After about 5 seconds from successfully signing in. Not sure why this happens.

morganick commented on docs/recipes/LocalFirstDataWithPowerSync.md:
```diff
@@ -0,0 +1,2037 @@
+
+#### Add Babel Plugin
+
+Update `babel.config.js` to include the `transform-async-generator-functions` plugin:
```
Add dev dependency as well `@babel/plugin-transform-async-generator-functions`

morganick reviewed (changes requested):
missing dev dep.
